jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java

```diff
@@ -411,53 +412,87 @@ public void complete(Callback callback)
         ByteBuffer content = null;
         synchronized (_channelState)
         {
-            switch (_state)
+            // First check the API state for any unrecoverable situations
+            switch (_apiState)
             {
-                case CLOSED:
-                    succeeded = true;
+                case UNREADY: // isReady() has returned false so a call to onWritePossible may happen at any time
+                    error = new CancellationException("Completed whilst write unready");
                     break;
 
-                case CLOSE:
-                case CLOSING:
-                    _closedCallback = Callback.combine(_closedCallback, callback);
+                case PENDING: // an async write is pending and may complete at any time
+                    // If this is not the last write, then we must abort
+                    if (!_channel.getResponse().isContentComplete(_written))
+                        error = new CancellationException("Completed whilst write pending");
                     break;
 
-                case OPEN:
-                    if (_onError != null)
-                    {
-                        error = _onError;
-                        break;
-                    }
+                case BLOCKED: // another thread is blocked in a write or a close
+                    error = new CancellationException("Completed whilst write blocked");
+                    break;
 
-                    _closedCallback = Callback.combine(_closedCallback, callback);
+                default:
+                    break;
+            }
 
-                    switch (_apiState)
-                    {
-                        case BLOCKING:
-                            // Output is idle blocking state, but we still do an async close
-                            _apiState = ApiState.BLOCKED;
-                            _state = State.CLOSING;
-                            content = BufferUtil.hasContent(_aggregate) ? _aggregate : BufferUtil.EMPTY_BUFFER;
-                            break;
+            // If we can't complete due to the API state, then abort
+            if (error != null)
+            {
+                _writeBlocker.fail(error);
+                _channel.abort(error);
+                _state = State.CLOSED;
```

Comment:
I'm not sure about the order of these calls.
`_writeBlocker.fail(error)` wakes up an application thread, so should we not abort and change the state before waking up threads?